docs/design.md

```diff
@@ -967,6 +864,92 @@ spare capacity is chosen in the same datacenter and a special-case split
 is done which simply duplicates the data 1:1 and resets the range
 configuration metadata.
 
+# Raft - Consistency of Range Replicas
+
+Each range is configured to consist of three or more replicas, as specified by
+their ZoneConfig. The replicas in a range maintain their own instance of a
+distributed consensus algorithm. We use the [*Raft consensus algorithm*](https://ramcloud.stanford.edu/wiki/download/attachments/11370504/raft.pdf)
+as it is simpler to reason about and includes a reference implementation
+covering important details. Every write to replicas is logged twice.
+Once to RocksDB’s internal log and once to levedb itself as part of the
+Raft consensus log.
+[ePaxos](https://www.cs.cmu.edu/~dga/papers/epaxos-sosp2013.pdf) has
+promising performance characteristics for WAN-distributed replicas, but
+it does not guarantee a consistent ordering between replicas.
+
+Raft elects a relatively long-lived leader which must be involved to
+propose commands. It heartbeats followers periodically and keeps their logs
+replicated. In the absence of heartbeats, followers become candidates
+after randomized election timeouts and proceed to hold new leader
+elections. Cockroach weights random timeouts such that the replicas with
+shorter round trip times to peers are more likely to hold elections
+first (not implemented yet). Only the Raft leader may propose commands;
+followers will simply relay commands to the last known leader.
+
+Our Raft implementation was developed together with CoreOS, but adds an extra
+layer of optimization to account for the fact that a single Node may have
+millions of consensus groups (one for each Range). Areas of optimization
+are chiefly coalesced heartbeats (so that the number of nodes dictates the
+number of heartbeats as opposed to the much larger number of ranges) and
+batch processing of requests.
+Future optimizations may include two-phase elections and quiescent ranges
+(i.e. stopping traffic completely for inactive ranges).
+
+# Range Leadership (Leader Leases)
+
+As outlined in the Raft section, the replicas of a Range are organized as
+a Raft group and execute commands from their shared commit log. Going
+through Raft is an expensive operation though, and there are tasks which
```

Comment:
No, it's just performance. If we sent all read operations through the raft log then we could do away with the concept of leader leases. This would be prohibitively expensive, however: all reads would incur the latency of a quorum network operation and writes to multiple disks.